### PR TITLE
50-Execute-Complex-message-part 1

### PR DESCRIPTION
**Pull request by DarjaLeonova**

### CODE REVIEW

File: template/template.go [hunk 1]
```diff
@@ -93,47 +94,114 @@ func (e *executer) execute() error {
...
+	return nil
 }
 
-func (e *executer) resolveSimpleMessage(message ast.SimpleMessage) error {
-	for _, pattern := range message {
-		switch pattern := pattern.(type) {
-		case ast.TextPattern:
-			if err := e.write(string(pattern)); err != nil {
-				return err
+func (e *executer) resolveDeclarations(declarations []ast.Declaration) error {
```
Comment (VladislavsPerkanuks) [inline template/template.go:121]: please add duplicate declaration error handling, and test for it
https://github.com/unicode-org/message-format-wg/blob/122e64c2482b54b6eff4563120915e0f86de8e4d/spec/errors.md#duplicate-declaration

File: template/template.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ var (
 	ErrDuplicateOptionName   = errors.New("duplicate option name")
 	ErrUnsupportedExpression = errors.New("unsupported expression")
 	ErrFormatting            = errors.New("formatting error")
+	ErrUnsupportedStatement  = errors.New("unsupported statement")
```
Comment (VladislavsPerkanuks) [inline template/template.go:22]: Add test for it in Test_ExecuteErrors

File: template/template.go [hunk 3]
```diff
@@ -93,47 +94,114 @@ func (e *executer) execute() error {
...
+			return fmt.Errorf("resolve expression: %w", err)
+		}
+
+		if err := e.write(resolved); err != nil {
+			return fmt.Errorf("resolve expression: %w", err)
 		}
+	case ast.Markup:
+		if err := e.write(fmt.Sprintf("%v", patternType)); err != nil {
+			return fmt.Errorf("write markup pattern: %w", err)
+		}
```
Comment (VladislavsPerkanuks) [inline template/template.go:180]: need to resolve markup before writing. 
Not this PR

File: template/template.go [hunk 4]
```diff
@@ -93,47 +94,114 @@ func (e *executer) execute() error {
...
+	default:
+		return fmt.Errorf("unknown pattern type: %T", patternType)
```
Comment (VladislavsPerkanuks) [inline template/template.go:182]: impossible case

**Review by janishorsts — APPROVED** [review]

File: template/template.go [hunk 5]
```diff
@@ -93,47 +95,124 @@ func (e *executer) execute() error {
...
-			if err := e.resolveExpression(pattern); err != nil {
+
+			e.input[string(d.Variable)] = resolved
+
+		case ast.InputDeclaration:
+			if _, ok := m[d.Operand.(ast.Variable)]; ok {
+				return fmt.Errorf("%w '%s'", ErrDuplicateDeclaration, "duplicate declaration")
+			}
+
+			m[d.Operand.(ast.Variable)] = struct{}{} //nolint: forcetypeassert
```
Comment (VladislavsPerkanuks) [inline template/template.go:148]:
```suggestion
			m[d.Operand] = struct{}{}
```

File: template/template.go [hunk 6]
```diff
@@ -93,47 +95,124 @@ func (e *executer) execute() error {
...
+			if err != nil {
+				return fmt.Errorf("resolve expression: %w", err)
 			}
-		case ast.Expression:
-			if err := e.resolveExpression(pattern); err != nil {
+
+			e.input[string(d.Variable)] = resolved
+
+		case ast.InputDeclaration:
+			if _, ok := m[d.Operand.(ast.Variable)]; ok {
```
Comment (VladislavsPerkanuks) [inline template/template.go:144]:
```suggestion
			if _, ok := m[d.Operand]; ok {
```

File: template/template_test.go [hunk 7]
```diff
@@ -159,70 +159,197 @@ func Test_ExecuteSimpleMessage(t *testing.T) {
...
+	type args struct {
+		inputMap map[string]any
+		inputStr string
+	}
+
 	tests := []struct {
 		name               string
 		expectedParseErr   error
 		expectedExecuteErr error
 		inputStr           string
```
Comment (VladislavsPerkanuks) [inline template/template_test.go:253]:
```suggestion

```

**Review by VladislavsPerkanuks — APPROVED** [review]

**Pull request merged**